feat(stremio): add presence

DarkVillager opened the pull request:
**Describe the changes in this pull request:**
_Added support for streamio with:_
* settings,
* buttons,
* timestamps,
* etc;
<details>
<summary> Proof showing the creation/modification is working as expected </summary>
<br>
<!-- Required when a presence has been added or modified --->
<!-- Attach screenshot(s) here --->
<img src="https://i.imgur.com/8Uy5nZC.png">
<img src="https://i.imgur.com/Nl5ul8H.png">
<img src="https://i.imgur.com/c3btNWN.png">
<img src="https://i.imgur.com/6NlGNNc.png">
<img src="https://i.imgur.com/eiF5xyn.png">
<img src="https://i.imgur.com/82slrGw.png">
<img src="https://i.imgur.com/POC0J23.png">
<img src="https://i.imgur.com/XRETFBa.png">
<img src="https://i.imgur.com/NgoUHRc.png">
<img src="https://i.imgur.com/uJlEiJV.png">
<img src="https://i.imgur.com/PvVQSab.png">
<img src="https://i.imgur.com/kuO9PNG.png">
<img src="https://i.imgur.com/5bS4CTM.png">
<img src="https://i.imgur.com/ybi4HNV.png">
<img src="https://i.imgur.com/TvGRDiu.png">
<img src="https://i.imgur.com/IZbMVP4.png">
<img src="https://i.imgur.com/rp7zaLY.png">
<img src="https://i.imgur.com/SR07zA1.png">
<img src="https://i.imgur.com/juXfZTF.png">
<img src="https://i.imgur.com/vxCZ2e5.png">
</details>

DarkVillager committed:
Save 1

DarkVillager committed:
feat(streamio): add presence

- Added support for streamio with:
* settings,
* buttons,
* timestamps,
* etc;

DarkVillager committed:
Deepscan Issue Fixed

Slowlife01 commented on websites/S/Stremio/presence.ts:
```diff
@@ -0,0 +1,148 @@
+				!pauseCheck &&
+				!document.querySelector("#loading-logo").className.includes("flashing")
+			) {
+				presenceData.endTimestamp = timestamp[1] ?? 0;
```
I don't think it will be undefined?
```suggestion
				presenceData.endTimestamp = timestamp[1];
```

Slowlife01 commented on websites/S/Stremio/presence.ts:
```diff
@@ -0,0 +1,148 @@
+							"[class='ng-binding ng-scope selected']"
+						)[2]?.textContent ?? "None";
+				}
+				if (genreSort == "Top") genreSort = "All";
```
```suggestion
				if (genreSort === "Top") genreSort = "All";
```

Slowlife01 commented on websites/S/Stremio/presence.ts:
```diff
@@ -0,0 +1,148 @@
+						.querySelector(
+							"#detail > div.details-less-info > div.details-top > div:nth-child(1)"
+						)
+						?.firstElementChild?.getAttribute("src") ?? "logo";
```
```suggestion
						?.firstElementChild.getAttribute("src") ?? "logo";
```

Slowlife01 commented on websites/S/Stremio/presence.ts:
```diff
@@ -0,0 +1,148 @@
+		video = document.querySelector<HTMLMediaElement>("#videoPlayer");
+		if (privacy && !video) presenceData.details = "Browsing...";
+		else if (privacy && video) presenceData.details = "Watching...";
+		else if (search && search.value) {
```
```suggestion
		else if (search?.value) {
```

Slowlife01 commented on websites/S/Stremio/presence.ts:
```diff
@@ -0,0 +1,148 @@
+			search = document.querySelector(
+				"#addons > div.filter > form > div.addon-search-input > input"
+			);
+			if (search && search.value) {
```
```suggestion
			if (search?.value) {
```

Slowlife01 commented on websites/S/Stremio/presence.ts:
```diff
@@ -0,0 +1,148 @@
+			presenceData.details = document
+				.querySelector("head > title")
+				?.textContent.replace("Stremio -", "");
+		} else if (window.location.pathname == "/")
```
```suggestion
		} else if (window.location.pathname === "/")
```

Slowlife01 commented on websites/S/Stremio/presence.ts:
```diff
@@ -0,0 +1,148 @@
+				const split = document
+					.querySelector("#play-progress-text")
+					?.textContent.split("/");
+				if (split && split[0]) {
```
```suggestion
				if (split?.[0]) {
```

Slowlife01 commented on websites/S/Stremio/presence.ts:
```diff
@@ -0,0 +1,148 @@
+			else presenceData.details = "Library";
+		} else if (page.includes("/calendar")) presenceData.details = "Calendar";
+		else if (page.includes("player")) {
+			if (video?.duration && video) {
```
```suggestion
			if (video?.duration) {
```

Slowlife01 commented on websites/S/Stremio/dist/metadata.json:
```diff
@@ -0,0 +1,45 @@
+			"if": {
+				"privacy": false
+			},
+			"title": "Thumbnails Mode",
```
```suggestion
			"title": "Show Thumbnail",
```

Slowlife01 reviewed (changes requested):
You don't have to keep updating your branch

DarkVillager committed:
Update websites/S/Stremio/dist/metadata.json

Co-authored-by: Slowlife <slowlife1165@gmail.com>

DarkVillager commented:
> You don't have to keep updating your branch

ok

DarkVillager committed:
Made Requested Changes.

DarkVillager committed:
Made button changes.

DarkVillager committed:
Revert "Made button changes."

This reverts commit 3ce3cab23f1dae5fa5b75bbb8aad43a089bb2a24.

DarkVillager committed:
right brach, button changes.

DarkVillager committed:
Dont use ` without variables

ImRodry commented on websites/S/Stremio/presence.ts:
```diff
@@ -0,0 +1,201 @@
+				if (active) {
+					presenceData.state = `${active.charAt(0).toUpperCase()}${active.slice(
+						1,
+						active.length
```
this is unnecessary

DarkVillager replied:
Ok, but no capital at start then.

ImRodry commented on websites/S/Stremio/presence.ts:
```diff
@@ -0,0 +1,201 @@
+				if (buttons) {
+					presenceData.buttons = [
+						{
+							label: `Browse ${title}`,
```
same as above, don't mention the title twice

ImRodry commented on websites/S/Stremio/presence.ts:
```diff
@@ -0,0 +1,201 @@
+				.trim();
+			if (genreSort) {
+				presenceData.state = `Sorted by: ${genreSort.substring(0, 1)}${genreSort
+					.substring(1, genreSort.length)
```
```suggestion
					.slice(1)
```

ImRodry commented on websites/S/Stremio/presence.ts:
```diff
@@ -0,0 +1,201 @@
+				.replace("_", " ")
+				.trim();
+			if (genreSort) {
+				presenceData.state = `Sorted by: ${genreSort.substring(0, 1)}${genreSort
```
```suggestion
				presenceData.state = `Sorted by: ${genreSort.charAt(0)}${genreSort
```

ImRodry commented on websites/S/Stremio/presence.ts:
```diff
@@ -0,0 +1,201 @@
+				if (buttons) {
+					presenceData.buttons = [
+						{
+							label: `View ${active} Library`,
```
same as above, don't repeat the name

ImRodry commented on websites/S/Stremio/presence.ts:
```diff
@@ -0,0 +1,201 @@
+			if (buttons) {
+				presenceData.buttons = [
+					{
+						label: `Join View Party For: ${title}`,
```
no colon after for, also don't repeat the title
```suggestion
						label: "Join View Party",
```

ImRodry reviewed (changes requested):
Also overall I think you should avoid checking `if (buttons)` every time you wanna set the button and just delete them at the end if buttons is false

DarkVillager commented on websites/S/Stremio/presence.ts:
```diff
@@ -0,0 +1,201 @@
+		clientId: "969208766807547917"
+	}),
+	browsingTimestamp = Math.floor(Date.now() / 1000);
+let video: HTMLMediaElement, timestamp: [number, number], pauseCheck: boolean;
```
This is prettier doing prettier things.

ImRodry replied:
I'm talking about empty lines separating chunks of code like one before or after this let statement, not in between lets

DarkVillager committed:
Made requested changes.

DarkVillager committed:
Lint Error

ImRodry commented on websites/S/Stremio/presence.ts:
```diff
@@ -0,0 +1,176 @@
+				presenceData.details = "Searching Addons For:";
+				presenceData.state = search.value;
+			} else {
+				if (active) presenceData.state = active;
+				else presenceData.state = "All";
```
```suggestion
				presenceData.state = active ?? "All";
```

ImRodry commented on websites/S/Stremio/presence.ts:
```diff
@@ -0,0 +1,176 @@
+
+				presenceData.buttons = [
+					{
+						label: "Browse",
```
Browse? Browse what?

ImRodry commented on websites/S/Stremio/presence.ts:
```diff
@@ -0,0 +1,176 @@
+				if (genreSort === "Top") genreSort = "All";
+				presenceData.buttons = [
+					{
+						label: "Browse",
```
ditto

ImRodry commented on websites/S/Stremio/presence.ts:
```diff
@@ -0,0 +1,176 @@
+						presence.timestampFromFormat(split[0].trim()),
+						presence.timestampFromFormat(split[1].trim())
+					);
+				} else timestamp = void 0;
```
timestamp is not assigned a value prior to this, this is not needed

DarkVillager committed:
Made Requested Changes

ImRodry commented on websites/S/Stremio/presence.ts:
```diff
@@ -0,0 +1,175 @@
+				presenceData.buttons = [
+					{
+						label: `Browse ${active}`,
+						url: document.location.href
```
```suggestion
						url: page
```

ImRodry commented on websites/S/Stremio/presence.ts:
```diff
@@ -0,0 +1,175 @@
+			presenceData.buttons = [
+				{
+					label: "View Library",
+					url: document.location.href
```
```suggestion
					url: page
```

ImRodry commented on websites/S/Stremio/presence.ts:
```diff
@@ -0,0 +1,175 @@
+			presenceData.buttons = [
+				{
+					label: "View Calendar",
+					url: document.location.href
```
```suggestion
					url: page
```

ImRodry commented on websites/S/Stremio/presence.ts:
```diff
@@ -0,0 +1,175 @@
+			presenceData.buttons = [
+				{
+					label: "Join View Party",
+					url: document.location.href
```
```suggestion
					url: page
```

ImRodry reviewed (changes requested):
Please don't forget to address the remaining reviews

DarkVillager committed:
Update websites/S/Stremio/dist/metadata.json

Co-authored-by: Rodry <38259440+ImRodry@users.noreply.github.com>

DarkVillager committed:
Update websites/S/Stremio/dist/metadata.json

Co-authored-by: Rodry <38259440+ImRodry@users.noreply.github.com>

DarkVillager committed:
Combined let

DarkVillager committed:
Fix browsingTimestamp

I was stupid & tired

Requested Changes

Update websites/S/Stremio/presence.ts

Update websites/S/Stremio/presence.ts

Apply suggestions from code review

Prettier

My Dutch Descriptions Sucks

Update presence.ts

browsingTimestamp

no vars

Co-Authored-By: Rodry <38259440+ImRodry@users.noreply.github.com>
Co-Authored-By: Slowlife <slowlife1165@gmail.com>
Co-Authored-By: QkeleQ10 <quinten@althues.nl>

ImRodry reviewed (changes requested):
You should address the remaining unresolved conversations

ImRodry commented on websites/S/Stremio/presence.ts:
```diff
@@ -0,0 +1,179 @@
+				if (genreSort === "Top") genreSort = "All";
+				presenceData.buttons = [
+					{
+						label: `Browse ${active}`,
```
Can you show an example of this?

DarkVillager replied:
![image](https://user-images.githubusercontent.com/42322979/169229863-165976ba-46f8-47a2-be1b-117cc05fa56f.png)

DarkVillager committed:
Revert "Merge branch 'main' into streamio"

This reverts commit c77695cab57dd922486546a34226c5e48b677836, reversing
changes made to 06069ba2a7c84b737d4ac623e75d5d05a27c65fc.

Revert "Revert "Merge branch 'main' into streamio""

This reverts commit 6d7f5404aa61134b9cfa13185f763e65f5945829.

DarkVillager commented on websites/S/Stremio/presence.ts:
```diff
@@ -0,0 +1,175 @@
+		video = document.querySelector<HTMLMediaElement>("#videoPlayer");
+		if (privacy && !video) presenceData.details = "Browsing...";
+		else if (privacy && video) presenceData.details = "Watching...";
+		else if (search?.value) {
+			presenceData.details = "Searching For:";
+			presenceData.state = search.value;
```
If privacy false display search results is okay right?

ImRodry replied:
yes this was resolved

EncryptedDev commented on websites/S/Stremio/presence.ts:
```diff
@@ -0,0 +1,179 @@
+	}),
+	browsingTimestamp = Math.floor(Date.now() / 1000);
+
+let video: HTMLMediaElement,
+	timestamp: [number, number],
+	pauseCheck: boolean,
+	search: HTMLInputElement,
+	genreSort: string,
+	title: string;
```
No

EncryptedDev replied:
Most of these variables don't get their value reassigned so use a global variable instead

DarkVillager replied:
Only video doesnt get reassigned

EncryptedDev commented on websites/S/Stremio/presence.ts:
```diff
@@ -0,0 +1,179 @@
+			largeImageKey: "logo",
+			startTimestamp: browsingTimestamp
+		},
+		page = window.location.href,
```
```suggestion
		{ pathname } = document.location,
```

DarkVillager replied:
maybe tell ur colleague to stop requesting if u want differently

EncryptedDev replied:
I don't mind how you do it, just destructure the object instead of making another var name for consistency

EncryptedDev commented on websites/S/Stremio/presence.ts:
```diff
@@ -0,0 +1,179 @@
+			presenceData.buttons = [
+				{
+					label: "Watch Video",
+					url: page
```
```suggestion
					url: document.location.href
```

EncryptedDev commented on websites/S/Stremio/presence.ts:
```diff
@@ -0,0 +1,179 @@
+				presenceData.buttons = [
+					{
+						label: "Browse Addons",
+						url: page
```
```suggestion
					url: document.location.href
```

EncryptedDev commented on websites/S/Stremio/presence.ts:
```diff
@@ -0,0 +1,179 @@
+				"#addons > div.filter > form > div.addon-search-input > input"
+			);
+			if (search?.value) {
+				presenceData.details = "Searching Addons For:";
```
```suggestion
				presenceData.details = "Searching addons for";
```

EncryptedDev commented on websites/S/Stremio/presence.ts:
```diff
@@ -0,0 +1,179 @@
+						url: page
+					}
+				];
+				presenceData.details = `Browsing ${title}:`;
```
```suggestion
				presenceData.details = `Browsing ${title}`;
```

EncryptedDev commented on websites/S/Stremio/presence.ts:
```diff
@@ -0,0 +1,179 @@
+			presenceData.details = `${
+				document.querySelector("[class='ng-scope ng-binding active']")
+					?.textContent ?? "General"
+			} Settings`;
```
```suggestion
			} settings`;
```

EncryptedDev commented on websites/S/Stremio/presence.ts:
```diff
@@ -0,0 +1,179 @@
+			presenceData.buttons = [
+				{
+					label: "View Library",
+					url: page
```
```suggestion
					url: document.location.href
```

EncryptedDev commented on websites/S/Stremio/presence.ts:
```diff
@@ -0,0 +1,179 @@
+			presenceData.buttons = [
+				{
+					label: "View Calendar",
+					url: page
```
```suggestion
					url: document.location.href
```

EncryptedDev commented on websites/S/Stremio/presence.ts:
```diff
@@ -0,0 +1,179 @@
+				}
+			];
+		} else if (window.location.pathname === "/")
+			presenceData.details = "Homepage";
```
```suggestion
			presenceData.details = "Vieiwng home page";
```

EncryptedDev commented on websites/S/Stremio/presence.ts:
```diff
@@ -0,0 +1,179 @@
+			presenceData.buttons = [
+				{
+					label: "Join View Party",
+					url: page
```
```suggestion
					url: document.location.href
```

DarkVillager committed:
fix(Stremio): pathname, href, acces and hostname

fix(Stremio): access if null

fix(Stremio): href

fix(Stremio): hostname

DarkVillager committed:
Update websites/S/Stremio/presence.ts

Co-authored-by: Rhys <64903135+EncryptedDev@users.noreply.github.com>

ImRodry commented on websites/S/Stremio/dist/metadata.json:
```diff
@@ -0,0 +1,51 @@
+	"service": "Stremio",
+	"description": {
+		"nl": "Vrijheid om te streamen. Stremio is een modern, alles-in-één mediacentrum voor video-entertainment. Ontdek, bekijk en organiseer video-inhoud via de eenvoudig te installeren invoegtoepassingen.",
+		"en": "Freedom To Stream. Stremio is a modern media center that's a one-stop solution for your video entertainment. Discover, watch and organize video content from easy to install addons."
```
just do me a favor and move this above the nl translation

ImRodry commented on websites/S/Stremio/presence.ts:
```diff
@@ -0,0 +1,177 @@
+				genreSort = document.querySelector(
+					"[class='ng-scope ng-binding selected']"
+				)?.textContent;
+				if (!genreSort) {
+					genreSort =
```
Use logical assignment

ImRodry commented on websites/S/Stremio/presence.ts:
```diff
@@ -0,0 +1,177 @@
+					.querySelector("#play-progress-text")
+					?.textContent.split("/");
+				if (split?.[0]) {
+					timestamp = presence.getTimestamps(
+						presence.timestampFromFormat(split[0].trim()),
+						presence.timestampFromFormat(split[1].trim())
+					);
```
please test this but I believe you could simplify this code
```suggestion
					timestamp = presence.timestampFromFormat(...split.map(s => s.trim()));
```

ImRodry commented on websites/S/Stremio/presence.ts:
```diff
@@ -0,0 +1,177 @@
+				}
+			];
+		} else if (pathname === "/") presenceData.details = "Viewing the homepage";
+	} else presenceData.details = "Browsing...";
```
so you have nothing for streamio.com?

DarkVillager replied:
yup just browsing.

EncryptedDev replied:
Could you not add anything for it?

DarkVillager replied:
> Could you not add anything for it?

sure

DarkVillager committed:
fix(Stremio): fix pathname and timestamp shorter

DarkVillager commented on websites/S/Stremio/presence.ts:
```diff
@@ -0,0 +1,176 @@
+		else if (search?.value) {
+			presenceData.details = "Searching For:";
+			presenceData.state = search.value;
+		} else if (href.includes("/detail/")) {
```
![Screenshot_41](https://user-images.githubusercontent.com/42322979/172128709-f84ccfe3-e4ff-47e0-9feb-5cf6220a2348.jpg)

DarkVillager commented:
Reason for href use instead of pathname
![Screenshot_41](https://user-images.githubusercontent.com/42322979/172128709-f84ccfe3-e4ff-47e0-9feb-5cf6220a2348.jpg)

EncryptedDev commented on websites/S/Stremio/presence.ts:
```diff
@@ -0,0 +1,176 @@
+						url: href
+					}
+				];
+				presenceData.details = `Browsing ${active}`;
+				presenceData.state = `Genre: ${genreSort}`;
```
Bit inconsistent 
```suggestion
				presenceData.details = `Browsing: ${active}`;
				presenceData.state = `Genre: ${genreSort}`;
```
```suggestion
				presenceData.details = `Browsing ${active}`;
				presenceData.state = `Genre ${genreSort}`;
```

EncryptedDev commented on websites/S/Stremio/presence.ts:
```diff
@@ -0,0 +1,176 @@
+				presenceData.state = `Genre: ${genreSort}`;
+			} else presenceData.state = "Browsing Movies";
+		} else if (href.includes("/library")) {
+			genreSort = document
+				.querySelector<HTMLSelectElement>(
+					"#library > div.sort-filter > div.custom-select.lib-sort > select"
+				)
+				?.value.replace("SORT", "")
+				.replace("_", " ")
+				.trim();
+			if (genreSort) {
+				presenceData.state = `Sorted by: ${genreSort.charAt(0)}${genreSort
+					.slice(1)
+					.toLowerCase()}`;
+			}
```
This seems quite useless, put something else here or nothing at all

DarkVillager committed:
Update websites/S/Stremio/presence.ts

Co-authored-by: Rhys <64903135+EncryptedDev@users.noreply.github.com>

DarkVillager committed:
feat(Stremio): add stremio.com

DarkVillager committed:
fix(Stremio): fix lint

ImRodry committed:
style: fix formatting issues

ImRodry commented:
Rebased to apply the latest changes in the prettier configuration. If you need help with your local branch, please ping me on the #presence-dev channel on Discord.

ImRodry reviewed (changes requested):
Just a final nit

DarkVillager commented on websites/S/Stremio/presence.ts:
```diff
@@ -0,0 +1,174 @@
+			largeImageKey: "logo",
+			startTimestamp: browsingTimestamp,
+		},
+		{ hostname, href } = document.location,
```
> All of the instances where href is used can be replaced by `pathname` which is more reliable. Obviously you should keep it on button URLs but for the other instances please replace it.

https://github.com/PreMiD/Presences/pull/6152#issuecomment-1147220707

ImRodry replied:
then you can use hash instead. Everything that comes after the # is a hash

ImRodry committed:
Merge branch 'main' into streamio

ImRodry commented:
Please address the failing CI and the pending reviews

DarkVillager committed:
fix(Stremio): lint and reviews

DarkVillager committed:
fix(Stremio): wrong position of the slash

DarkVillager committed:
fix(Stremio): ran prettier in terminal
